### PR TITLE
Spring category

### PR DESCRIPTION
**Pull request by TarasYurkovskyi**

### CODE REVIEW

**Review by Andriy-Hrynevetskyi — APPROVED** [review]
gj)

File: src/main/java/com/example/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+@RestController
+@RequestMapping(value = "/api/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @Operation(summary = "Create a new category",
+            description = "Create a new category with the provided information.")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PostMapping
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/CategoryController.java:37]:
```suggestion
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
```

File: src/main/java/com/example/bookstore/model/Category.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/Category.java:27]:
```suggestion
    @Column
    private String description;
```

File: src/main/java/com/example/bookstore/model/Category.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/Category.java:14]:
```suggestion
@Entity
@NoArgsConstructor
```

File: src/main/resources/db/changelog/changes/12-create-categories-table.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-categories-table
+      author: Taras
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/12-create-categories-table.yaml]:
```suggestion
      author: Taras Yurkovskiy
```

File: src/main/resources/db/changelog/changes/13-insert-categories-to-categories-table.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-categories-to-categories-table
+      author: Taras
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/13-insert-categories-to-categories-table.yaml]:
```suggestion
      author: Taras Yurkovskiy
```

File: src/main/resources/db/changelog/changes/14-create-books_categories-table.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-books_categories-table
+      author: Taras
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/14-create-books_categories-table.yaml]:
```suggestion
      author: Taras Yurkovskiy
```

File: src/main/resources/application.properties [hunk 7]
```diff
@@ -1,7 +1,7 @@
 spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 spring.datasource.url=jdbc:mysql://localhost:3306/jdbc
 spring.datasource.username=*
-spring.datasource.password=*
+spring.datasource.password=*!
```
Comment (d1sam) [inline src/main/resources/application.properties]:
```suggestion
spring.datasource.password=*
```

Comment (d1sam) [reply src/main/resources/application.properties]: it's unnecessary change

File: src/main/resources/db/changelog/changes/12-create-categories-table.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        - createTable:
+            tableName: categories
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/12-create-categories-table.yaml]: you have primary key, that's why nullable constraint is unnecessary
```suggestion
```

File: src/main/resources/db/changelog/changes/14-create-books_categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+      changes:
+        - createTable:
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_books_categories_book_id
+                    nullable: false
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/14-create-books_categories-table.yaml]: you have primary key, that's why nullable constraint is unnecessary
```suggestion
```

File: src/main/resources/db/changelog/changes/14-create-books_categories-table.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+                    references: books(id)
+                    deleteCascade: true
+                    primaryKey: true
+              - column:
+                  name: category_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_books_categories_category_id
+                    nullable: false
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/14-create-books_categories-table.yaml]: you have primary key, that's why nullable constraint is unnecessary
```suggestion
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but some changes should be done!

File: src/main/resources/application.properties (hunk 7 above)
Comment (TarasYurkovskyi) [reply src/main/resources/application.properties]: its missclick)

**Review by d1sam — APPROVED** [review]
well done!

**Pull request merged**